runtime error on google directory sync

## What happened?

Cache service crashes while syncing with google directory service after service start.

## What did you expect to happen?

Not to crash.

## What's your environment like?

- Pomerium version (retrieve with `pomerium --version`): 0.11.0
- Server Operating System/Architecture/Cloud: Google GKE

## What's your config.yaml?

```config.yaml
log_level: info
address: ":8443"
grpc_address: ":8443"
autocert: false
autocert_must_staple: false
certificate_file: "/pomerium/cert.pem"
certificate_key_file: "/pomerium/privkey.pem"
certificate_authority_file: "/pomerium/cert.pem"
forward_auth_url: "https://fwdauth.example.com"
authenticate_service_url: "https://auth.example.com"
authorize_service_url: "https://pomerium-authorize.pomerium.svc.cluster.local"
databroker_service_url: "https://pomerium-cache.pomerium.svc.cluster.local"
cookie_secret: *SCRUBBED*
shared_secret: *SCRUBBED*
signing_key: *SCRUBBED*
databroker_storage_type: "memory"
idp_provider: "google"
idp_provider_url: "https://accounts.google.com"
idp_client_id: *SCRUBBED*
idp_client_secret: *SCRUBBED*
idp_service_account: *SCRUBBED*
idp_refresh_directory_interval: "60m"
idp_refresh_directory_timeout: "15m"
policy:
    - from: https://whoami.example.com
      to: http://whoami.pomerium.svc.cluster.local
      allowed_groups:
        - team@example.com
```

## What did you see in the logs?

```logs
{"level":"info","service":"databroker","type":"type.googleapis.com/directory.User","id":"google/*SCRUBBED*","time":"2020-12-11T07:41:15Z","message":"set"}
{"level":"info","service":"databroker","type":"type.googleapis.com/directory.User","id":"google/*SCRUBBED*","time":"2020-12-11T07:41:15Z","message":"set"}
{"level":"info","service":"databroker","type":"type.googleapis.com/directory.User","id":"google/*SCRUBBED*","time":"2020-12-11T07:41:15Z","message":"set"}
{"level":"info","service":"databroker","type":"type.googleapis.com/directory.User","id":"google/*SCRUBBED*","time":"2020-12-11T07:41:15Z","message":"set"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xf5abe5]

goroutine 216 [running]:
github.com/google/btree.items.find(0xc00066e8c0, 0xc, 0xe, 0x18a7680, 0xc001563a00, 0xc, 0xf5f380)
	/home/runner/go/pkg/mod/github.com/google/btree@v1.0.0/btree.go:191 +0xc5
github.com/google/btree.(*node).iterate(0xc0015b8c40, 0x1, 0x18a7680, 0xc001563a00, 0x0, 0x0, 0xf50001, 0xc000627898, 0xc001563a00)
	/home/runner/go/pkg/mod/github.com/google/btree@v1.0.0/btree.go:508 +0x7c5
github.com/google/btree.(*node).iterate(0xc0029cd940, 0x1, 0x18a7680, 0xc001563a00, 0x0, 0x0, 0xc000620001, 0xc000627898, 0x80)
	/home/runner/go/pkg/mod/github.com/google/btree@v1.0.0/btree.go:529 +0x765
github.com/google/btree.(*node).iterate(0xc002091600, 0x1, 0x18a7680, 0xc001563a00, 0x0, 0x0, 0xc000620001, 0xc000627898, 0x1871491)
	/home/runner/go/pkg/mod/github.com/google/btree@v1.0.0/btree.go:529 +0x765
github.com/google/btree.(*BTree).AscendGreaterOrEqual(...)
	/home/runner/go/pkg/mod/github.com/google/btree@v1.0.0/btree.go:768
github.com/pomerium/pomerium/pkg/storage/inmemory.(*DB).List(0xc00042a700, 0x18cb580, 0xc000d09780, 0xc0008787f0, 0xc, 0xc0009d5108, 0xc0013b8dd0, 0xc0009d51b8, 0x0, 0x0)
	/home/runner/work/pomerium/pomerium/pkg/storage/inmemory/inmemory.go:133 +0x148
github.com/pomerium/pomerium/internal/databroker.(*Server).doSync(0xc00026f900, 0x18cb580, 0xc000d09780, 0x0, 0x0, 0xc000627a08, 0x18da600, 0xc00042a700, 0x18d7300, 0xc000fba7a0, ...)
	/home/runner/work/pomerium/pomerium/internal/databroker/server.go:305 +0x7c
github.com/pomerium/pomerium/internal/databroker.(*Server).Sync(0xc00026f900, 0xc000d44720, 0x18d7300, 0xc000fba7a0, 0x0, 0x0)
	/home/runner/work/pomerium/pomerium/internal/databroker/server.go:376 +0x510
github.com/pomerium/pomerium/cache.(*dataBrokerServer).Sync(0xc000b00200, 0xc000d44720, 0x18d7300, 0xc000fba7a0, 0xc000b00200, 0xc000a760c0)
	/home/runner/work/pomerium/pomerium/cache/databroker.go:94 +0x10b
github.com/pomerium/pomerium/pkg/grpc/databroker._DataBrokerService_Sync_Handler(0x160eb40, 0xc000b00200, 0x18d5800, 0xc000817d60, 0xc000817d60, 0x7ff71e32ee98)
	/home/runner/work/pomerium/pomerium/pkg/grpc/databroker/databroker.pb.go:1564 +0x10b
github.com/pomerium/pomerium/pkg/grpcutil.AttachMetadataInterceptors.func2(0x160eb40, 0xc000b00200, 0x18d5800, 0xc000817d60, 0xc000817d40, 0x174ea20, 0x174ea20, 0xc000817d60)
	/home/runner/work/pomerium/pomerium/pkg/grpcutil/version.go:22 +0x6b
google.golang.org/grpc.getChainStreamHandler.func1(0x160eb40, 0xc000b00200, 0x18d5800, 0xc000817d60, 0x13e5240, 0xc000fba770)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1332 +0xdd
github.com/pomerium/pomerium/internal/telemetry/requestid.StreamServerInterceptor.func1(0x160eb40, 0xc000b00200, 0x18d4480, 0xc000a760c0, 0xc000817d40, 0xc000d09680, 0xc000d09680, 0x1480601)
	/home/runner/work/pomerium/pomerium/internal/telemetry/requestid/grpc_server.go:30 +0x158
google.golang.org/grpc.chainStreamServerInterceptors.func1(0x160eb40, 0xc000b00200, 0x18d4480, 0xc000a760c0, 0xc000817d40, 0x174ea20, 0x18cb640, 0xc000fd8690)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1318 +0xbd
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0006de000, 0x18db440, 0xc000b2ef00, 0xc00096e100, 0xc0007476e0, 0x2321840, 0x0, 0x0, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1464 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc0006de000, 0x18db440, 0xc000b2ef00, 0xc00096e100, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1537 +0xc9c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000fb4a60, 0xc0006de000, 0x18db440, 0xc000b2ef00, 0xc00096e100)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:871 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:869 +0x1fd
```

## Additional context

Google directory contains several hundreds of users and groups. Groups are nested and may also contain groups from external companys. These external groups may not be resolveable.

--- a/pkg/storage/inmemory/inmemory.go
+++ b/pkg/storage/inmemory/inmemory.go
@@ -49,7 +49,7 @@ type DB struct {
 
 	lastVersion uint64
 
-	mu         sync.Mutex
+	mu         sync.RWMutex
 	byID       *btree.BTree
 	byVersion  *btree.BTree
 	deletedIDs []string
@@ -100,6 +100,9 @@ func (db *DB) Close() error {
 
 // Delete marks a record as deleted.
 func (db *DB) Delete(_ context.Context, id string) error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
 	defer db.onchange.Broadcast()
 	db.replaceOrInsert(id, func(record *databroker.Record) {
 		record.DeletedAt = ptypes.TimestampNow()
@@ -110,6 +113,9 @@ func (db *DB) Delete(_ context.Context, id string) error {
 
 // Get gets a record from the db.
 func (db *DB) Get(_ context.Context, id string) (*databroker.Record, error) {
+	db.mu.RLock()
+	defer db.mu.RUnlock()
+
 	record, ok := db.byID.Get(byIDRecord{Record: &databroker.Record{Id: id}}).(byIDRecord)
 	if !ok {
 		return nil, errors.New("not found")
@@ -119,6 +125,9 @@ func (db *DB) Get(_ context.Context, id string) (*databroker.Record, error) {
 
 // GetAll gets all the records in the db.
 func (db *DB) GetAll(_ context.Context) ([]*databroker.Record, error) {
+	db.mu.RLock()
+	defer db.mu.RUnlock()
+
 	var records []*databroker.Record
 	db.byID.Ascend(func(item btree.Item) bool {
 		records = append(records, item.(byIDRecord).Record)
@@ -129,6 +138,9 @@ func (db *DB) GetAll(_ context.Context) ([]*databroker.Record, error) {
 
 // List lists all the changes since the given version.
 func (db *DB) List(_ context.Context, sinceVersion string) ([]*databroker.Record, error) {
+	db.mu.RLock()
+	defer db.mu.RUnlock()
+
 	var records []*databroker.Record
 	db.byVersion.AscendGreaterOrEqual(byVersionRecord{Record: &databroker.Record{Version: sinceVersion}}, func(i btree.Item) bool {
 		record := i.(byVersionRecord)
@@ -142,6 +154,9 @@ func (db *DB) List(_ context.Context, sinceVersion string) ([]*databroker.Record
 
 // Put replaces or inserts a record in the db.
 func (db *DB) Put(_ context.Context, id string, data *anypb.Any) error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
 	defer db.onchange.Broadcast()
 	db.replaceOrInsert(id, func(record *databroker.Record) {
 		record.Data = data
@@ -165,9 +180,6 @@ func (db *DB) Watch(ctx context.Context) <-chan struct{} {
 }
 
 func (db *DB) replaceOrInsert(id string, f func(record *databroker.Record)) {
-	db.mu.Lock()
-	defer db.mu.Unlock()
-
 	record, ok := db.byID.Get(byIDRecord{Record: &databroker.Record{Id: id}}).(byIDRecord)
 	if ok {
 		db.byVersion.Delete(byVersionRecord(record))

--- a/pkg/storage/inmemory/inmemory_test.go
+++ b/pkg/storage/inmemory/inmemory_test.go
@@ -8,6 +8,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
 	"google.golang.org/protobuf/types/known/anypb"
 )
 
@@ -84,4 +85,24 @@ func TestDB(t *testing.T) {
 		require.Error(t, err)
 		assert.Nil(t, record)
 	})
+}
+
+func TestConcurrency(t *testing.T) {
+	ctx := context.Background()
+	db := NewDB("example", 2)
+
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		for i := 0; i < 1000; i++ {
+			_, _ = db.List(ctx, "")
+		}
+		return nil
+	})
+	eg.Go(func() error {
+		for i := 0; i < 1000; i++ {
+			db.Put(ctx, fmt.Sprint(i), new(anypb.Any))
+		}
+		return nil
+	})
+	eg.Wait()
 }

